modules/streams-core/src/main/scala/com.snowplowanalytics.snowplow/sources/internal/LowLevelSource.scala

```diff
@@ -344,4 +357,38 @@ private[sources] object LowLevelSource {
    */
   private def timeoutForFirstWindow(config: EventProcessingConfig.TimedWindows): FiniteDuration =
     (config.duration.toMillis * config.firstWindowScaling).toLong.milliseconds
+
+  private def batchUpCheckpoints[F[_]: Async, C](timeout: FiniteDuration, semigroup: Semigroup[C]): Pipe[F, C, C] = {
+
+    def go(timedPull: Pull.Timed[F, C], output: Option[C]): Pull[F, C, Unit] =
+      timedPull.uncons.flatMap {
+        case None =>
+          // Upstream finished cleanly. Emit whatever is pending and we're done.
+          Pull.outputOption1(output)
+        case Some((Left(_), next)) =>
+          // Timer timed-out. Emit whatever is pending.
+          Pull.outputOption1(output) >> go(next, None)
+        case Some((Right(chunk), next)) =>
+          // Upstream emitted tokens to us. We might already have pending tokens
+          output match {
+            case Some(c) =>
+              go(next, Some(chunk.foldLeft(c)(semigroup.combine(_, _))))
+            case None =>
+              semigroup.combineAllOption(chunk.iterator) match {
+                case Some(c) =>
+                  next.timeout(timeout) >> go(next, Some(c))
```

Comment:
If I understand correctly, you set the timeout here so that we start the timer only after we have received some `C` since last batch, is that right? That's smart!